### PR TITLE
Add "USER_NAME"  to BuildVariables

### PR DESCRIPTION
**Pull request by zhangzhenyu0819**
Add "USER_NAME"  to BuildVariables(ManualBuildCause only)

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 1]
```diff
@@ -46,6 +46,20 @@
...
+    public String getCauseUserName(AbstractBuild<?, ?> build) {
+        CauseAction causeAction = build.getAction(CauseAction.class);
+        List<Cause> buildCauses = causeAction.getCauses();
+
+        for (Cause cause : buildCauses) {
+            if (isUserCause(cause)){
+            	return getUserName(cause);
+            }
+        }
+
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: redundant line breaks

File: src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java [hunk 2]
```diff
@@ -1,24 +1,24 @@
 package org.jenkinsci.plugins.envinject;
 
+import static com.google.common.base.Joiner.*;
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: please dont use wildcard imports. They hamper to see which classes is actually imported

File: src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java [hunk 3]
```diff
@@ -39,13 +40,14 @@
     @SuppressWarnings("deprecation")
     @Test
     public void shouldWriteInfoAboutManualBuildCause() throws Exception {
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: please add test for not null value of username for user cause

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 4]
```diff
@@ -94,4 +108,21 @@ private static String getTriggerName(Cause cause) {
         return null;
     }
 
+    private static Boolean isUserCause(Cause cause) {
+    	if (Cause.UserIdCause.class.isInstance(cause)) {
+    		return true;
+        }
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: please format all changed code properly

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 5]
```diff
@@ -94,4 +108,21 @@ private static String getTriggerName(Cause cause) {
         return null;
     }
 
+    private static Boolean isUserCause(Cause cause) {
+    	if (Cause.UserIdCause.class.isInstance(cause)) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: This code should also support the deprecated <code>UserCause</code>, because it's still being used in Hudson-compatible plugins

Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: Also use <code>cause instanceof Cause.UserIdCause</code>

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 6]
```diff
@@ -94,4 +108,21 @@ private static String getTriggerName(Cause cause) {
...
+    		return true;
+        }
+
+    	return false;
+    }
+
+    private static String getUserName(Cause cause) {
+    	if (Cause.UserIdCause.class.isInstance(cause)) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: same as above (legacy class and Java operator)

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 7]
```diff
@@ -94,4 +108,21 @@ private static String getTriggerName(Cause cause) {
...
 
+    private static Boolean isUserCause(Cause cause) {
+    	if (Cause.UserIdCause.class.isInstance(cause)) {
+    		return true;
+        }
+
+    	return false;
+    }
+
+    private static String getUserName(Cause cause) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: Add <code>@CheckForNull</code>, because the method may return null

File: src/main/java/org/jenkinsci/plugins/envinject/service/EnvInjectVariableGetter.java [hunk 8]
```diff
@@ -3,14 +3,17 @@
 import hudson.EnvVars;
 import hudson.Util;
 import hudson.matrix.MatrixRun;
-import hudson.model.*;
+import hudson.model.Environment;
+import hudson.model.AbstractBuild;
+import hudson.model.Computer;
+import hudson.model.EnvironmentContributor;
+import hudson.model.Executor;
+import hudson.model.Hudson;
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/EnvInjectVariableGetter.java]: I agree with this refactoring, but it should be done in a separate pull-request

File: src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java [hunk 9]
```diff
@@ -39,13 +40,14 @@
...
-        Cause cause = Cause.UserCause.class.newInstance();
+    	UserIdCause cause = Cause.UserIdCause.class.newInstance();
```
Comment (oleg-nenashev) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: I just wonder how it may work, because you don't specify the username when you trigger the build... If the check below works, then something is wrong with the matcher

Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: @oleg-nenashev matcher works fine. Both values are null. So it equal

**Comment by zhangzhenyu0819** [comment]
@lanwen @oleg-nenashev Thank you for your comment.
I've modified the code.

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 10]
```diff
@@ -46,6 +48,32 @@
...
+        CauseAction causeAction = build.getAction(CauseAction.class);
+        List<Cause> buildCauses = causeAction.getCauses();
+
+        for (Cause cause : buildCauses) {
+            if (isUserCause(cause) || isUserIdCause(cause)) {
+            	return getUserName(cause);
+            }
+        }
+
+		return null;
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: reformat this please and rm prepending empty line

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 11]
```diff
@@ -94,4 +122,43 @@ private static String getTriggerName(Cause cause) {
...
+
+    	return false;
+    }
+
+    @CheckForNull
+    private static String getUserName(Cause cause) {
+    	if (cause instanceof UserIdCause) {
+    		Cause.UserIdCause userIdCause = (UserIdCause) cause;
+    		return userIdCause.getUserName();
+        }else if (cause instanceof UserCause) {
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: `} else`

File: src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java [hunk 12]
```diff
@@ -39,13 +41,19 @@
...
         FreeStyleBuild build = jenkins.createFreeStyleProject().scheduleBuild2(0, cause).get();
 
         assertThat(build.getResult(), is(SUCCESS));
         assertThat(build, withCause(BUILD_CAUSE, MANUAL_TRIGGER));
         assertThat(build, withCause(ROOT_BUILD_CAUSE, MANUAL_TRIGGER));
         assertThat(build, withCausesIsTrue(sub(BUILD_CAUSE, MANUAL_TRIGGER), sub(ROOT_BUILD_CAUSE, MANUAL_TRIGGER)));
+        if(cause.getUserName() != null){
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: no way. It should be separate test case

**Comment by lanwen** [comment]
What for you write both `user_id` and `user_name` as of it synonyms in such context?

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java (hunk 10 above)
Comment (zhangzhenyu0819) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: I have fix here.

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java (hunk 11 above)
Comment (zhangzhenyu0819) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: i'm sorry.. I have fix here.

File: src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java (hunk 12 above)
Comment (zhangzhenyu0819) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: added testcase.

**Comment by zhangzhenyu0819** [comment]
Additional items are being used in e-mail notification template.
e.g.
```
user_id:${USER_ID} user_name:${USER_NAME} Start this job.
```

File: src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java [hunk 13]
```diff
@@ -139,6 +142,32 @@ public String getShortDescription() {
...
+        assertThat(build.getResult(), is(SUCCESS));
+        assertThat(build, withCause(BUILD_CAUSE, MANUAL_TRIGGER));
+        assertThat(build, withCause(ROOT_BUILD_CAUSE, MANUAL_TRIGGER));
+        assertThat(build, withCausesIsTrue(sub(BUILD_CAUSE, MANUAL_TRIGGER), sub(ROOT_BUILD_CAUSE, MANUAL_TRIGGER)));
+        assertThat(build, withCause(USER_NAME, cause.getUserName()));
+    }
+
+    @Test
+    public void shouldWriteInfoUserIdCause() throws Exception {
+    	UserIdCause cause = Cause.UserIdCause.class.newInstance();
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: and where user_id is set to check it? You just copied prev testcase. Unit test should test only what you want to check. Please add separate cases for nonnull/null user_id and user_name only. Don't touch other cases if they works fine.

**Comment by lanwen** [comment]
@zhangzhenyu0819 is you need both user_id and user_name at same time? It adds a lot of duplication in this implementation.

**Comment by zhangzhenyu0819** [comment]
@lanwen I wishes to obtain all of the user information that can be achieved, including user_id and user_name.

File: src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java [hunk 14]
```diff
@@ -139,6 +142,25 @@ public String getShortDescription() {
         assertThat(build, withCause(ROOT_BUILD_CAUSE, ""));
     }
 
+    @SuppressWarnings("deprecation")
+    @Test
+    public void shouldWriteInfoUserCause() throws Exception {
+        UserCause cause = Cause.UserCause.class.newInstance();
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/BuildCauseRetrieverTest.java]: when you create cause in such way, username and user id is null. Please specify it explicitly

**Comment by zhangzhenyu0819** [comment]
@lanwen I have modified the test case, please check it.

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 15]
```diff
@@ -94,4 +118,39 @@ private static String getTriggerName(Cause cause) {
...
+        } else if (cause instanceof UserCause) {
+    		Cause.UserCause userCause = (UserCause) cause;
+    		return userCause.getUserName();
+        }
+        return null;
+    }
+
+    @CheckForNull
+    private static String getUserId(Cause cause) {
+    	if (cause instanceof UserIdCause) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: Missing UserCause handling

**Comment by lanwen** [comment]
@zhangzhenyu0819 thanks, now tests is ok

File: src/main/java/org/jenkinsci/plugins/envinject/service/EnvInjectVariableGetter.java [hunk 16]
```diff
@@ -111,6 +112,20 @@
         Map<String, String> triggerVariable = new BuildCauseRetriever().getTriggeredCause(build);
         result.putAll(triggerVariable);
 
+        //Get Cause UserName
+        String userName = new BuildCauseRetriever().getCauseUserName(build);
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/EnvInjectVariableGetter.java]: I would refactor the code to call <code>new BuildCauseRetriever()</code> only once in order to simplify the life of the garbage collector

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 17]
```diff
@@ -46,6 +48,28 @@
         return env;
     }
 
+    public String getCauseUserName(AbstractBuild<?, ?> build) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: The method needs to be annotated an documented  properly.
Also the parameter should be <code>Run<?,?></code> unless you need something from the Abstract build

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 18]
```diff
@@ -46,6 +48,28 @@
...
+        List<Cause> buildCauses = causeAction.getCauses();
+        for (Cause cause : buildCauses) {
+            if (isUserCause(cause) || isUserIdCause(cause)) {
+            	return getUserName(cause);
+            }
+        }
+		return null;
+	}
+
+    public String getCauseUserId(AbstractBuild<?, ?> build) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: Same

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 19]
```diff
@@ -94,4 +118,39 @@ private static String getTriggerName(Cause cause) {
         return null;
     }
 
+    private static Boolean isUserCause(Cause cause) {
+    	if (cause instanceof UserCause){
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: It is an obsolete cause. Should also check UserIdCause

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 20]
```diff
@@ -94,4 +118,39 @@ private static String getTriggerName(Cause cause) {
...
+    		return true;
+    	}
+    	return false;
+    }
+
+    private static Boolean isUserIdCause(Cause cause) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: Both calls should be in a single method IMHO

File: src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java [hunk 21]
```diff
@@ -1,8 +1,13 @@
 package org.jenkinsci.plugins.envinject.service;
 
+import static com.google.common.base.Joiner.*;
+import static org.apache.commons.lang.StringUtils.*;
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/envinject/service/BuildCauseRetriever.java]: Please do not use wildcards in the production code imports

**Comment by oleg-nenashev** [comment]
BTW, I still do not see much justification for this change since there is https://wiki.jenkins-ci.org/display/JENKINS/Build+User+Vars+Plugin

**Comment by zhangzhenyu0819** [comment]
sorry, I do oblivious to the build-user-vars-plugin.
Close this PR?
